framework/test/unit/modules/dpos_v2/auxiliary.spec.ts

```diff
@@ -0,0 +1,570 @@
+					EMPTY_KEY,
+					{ timestamp: previousTimestamp },
+					previousTimestampStoreSchema,
+				);
+
+				const missedBlocks: Record<string, number> = {};
+				for (let i = 0; i < 103; i += 1) {
+					// Generator address forged last block of previous round where
+					// previousTimestamp points out and also forging current end
+					// of round block. Because of that updateDelegateProductivity
+					// function will substract 2 from its missed blocks and hence
+					// we assign it 2, results in 0 meaning block is not forged yet.
+					missedBlocks[delegateAddresses[i].toString('hex')] = i === 102 ? 2 : 1;
+				}
+
+				when(validatorsAPI.getGeneratorsBetweenTimestamps)
+					.calledWith(context.getAPIContext(), previousTimestamp, currentTimestamp)
+					.mockReturnValue(missedBlocks);
+
+				when(validatorsAPI.getGeneratorAtTimestamp)
+					.calledWith(context.getAPIContext(), currentTimestamp)
+					.mockReturnValue(generatorAddress);
+
+				when(validatorsAPI.getGeneratorAtTimestamp)
+					.calledWith(context.getAPIContext(), previousTimestamp)
+					.mockReturnValue(generatorAddress);
+
+				await updateDelegateProductivity({
+					moduleConfig: defaultConfigs,
+					validatorsAPI,
+					context,
+					previousTimestamp,
+				});
+
+				expect.assertions(delegateAddresses.length + 1);
+				for (const delegateAddress of delegateAddresses) {
+					const currentDelegate = await delegateStore.getWithSchema<DelegateAccount>(
+						delegateAddress,
+						delegateStoreSchema,
+					);
+					if (delegateAddress.equals(generatorAddress)) {
+						expect(currentDelegate.consecutiveMissedBlocks).toBe(0);
+						expect(currentDelegate.lastGeneratedHeight).toBe(nextForgedHeight);
+					} else {
+						expect(currentDelegate.consecutiveMissedBlocks).toBe(1);
+					}
+				}
+			});
+		});
+
+		describe('When only 2 delegate missed a block since last block', () => {
+			it('should increment "consecutiveMissedBlocks" only for forgers who missed a block', async () => {
+				const lastForgerAddress = delegateAddresses[delegateAddresses.length - 4];
+				const generatorAddress = delegateAddresses[delegateAddresses.length - 1];
+				const missedForgers = [
+					delegateAddresses[delegateAddresses.length - 2],
+					delegateAddresses[delegateAddresses.length - 3],
+				];
+				const previousTimestamp = 10260;
+				const currentTimestamp = 10290;
+				const lastForgedHeight = 926;
+				const nextForgedHeight = lastForgedHeight + 1;
+
+				const context = createBlockContext({
+					header: {
+						height: nextForgedHeight,
+						generatorAddress,
+						timestamp: currentTimestamp,
+					} as any,
+					stateStore,
+				}).getBlockAfterExecuteContext();
+
+				await previousTimestampStore.setWithSchema(
+					EMPTY_KEY,
+					{ timestamp: previousTimestamp },
+					previousTimestampStoreSchema,
+				);
+
+				const missedBlocks: Record<string, number> = {};
+				for (let i = 0; i < missedForgers.length + 1 + 1; i += 1) {
```

Comment:
```suggestion
for (let i = 0; i < missedForgers.length + 2; i += 1) {
```

Reply:
I don't remember why I did that but it was about referring to different types of generators. If you still it's better that way then I'll change.

Reply:
I guess for tests its fine